llvm/lib/IR/VectorTypeUtils.cpp

```diff
@@ -27,6 +28,30 @@ Type *llvm::toVectorizedStructTy(StructType *StructTy, ElementCount EC) {
       }));
 }
 
+Type *llvm::toVectorizedRetTy(Type *Ty, ElementCount EC, unsigned IID) {
+  if (EC.isScalar())
+    return Ty;
+
+  if (StructType *StructTy = dyn_cast<StructType>(Ty)) {
+    assert(isUnpackedStructLiteral(StructTy) &&
+           "expected unpacked struct literal");
+    assert(all_of(StructTy->elements(), VectorType::isValidElementType) &&
+           "expected all element types to be valid vector element types");
+    return StructType::get(
+        StructTy->getContext(),
+        map_to_vector(enumerate(StructTy->elements()), [&](auto It) -> Type * {
+          Type *ElTy = It.value();
+          if (IID != 0 &&
+              isVectorIntrinsicWithStructReturnScalarAtField(IID, It.index()))
+            return ElTy;
+          return VectorType::get(ElTy, EC);
+        }));
```

Comment:
I think this code can still just go in `toVectorizedStructTy`, adding an `IID` argument (with a default of 0), rather than duplicating the same logic twice. This function can then just call `toVectorizedStructTy` with an `IID`.